Extend the IncludeGenerator interface with the Sling Request

santiagozky opened the pull request:
Extend the IncludeGenerator interface with the Sling Request to give more flexibility to potential implementations of the interface.

santiagozky commented:
The current interface is a bit limited. I want to create a custom IncludeGenerator that relies in some 'context' that could be obtained from the original request.

rombert commented:
Thanks for the PR @santiagozky . As this is a non-trivial addition, can you please file a Jira issue first under https://issues.apache.org/jira/browse/SLING so we can discuss the best way of implementing this?

santiagozky commented:
hi @rombert. Sure. I've created it https://issues.apache.org/jira/browse/SLING-10096

santiagozky commented:
sorry I never updated this. fell out of priority for me and havent have time to get around it.

santiagozky commented:
@rombert 
sorry it took long. I had to leave this for a bit.
It complains about one line without unit testing. do you need it?

rombert commented on src/main/java/org/apache/sling/dynamicinclude/Configuration.java:
```diff
@@ -66,11 +65,7 @@
-          @Option(label = "ESI", value = "ESI"),
-          @Option(label = "Javascript", value = "JSI")
-      })
+      @AttributeDefinition(name = "Include type", description = "Type of generated include tags")
```
Any reason for removing this?

santiagozky replied:
well, if the point is to allow extra implementations, that means we cannot longer have a fixed list of options. otherwise you cannot choose whatever you implement

rombert replied:
Makes sense. It might help if we listed the built-in implementations.

rombert commented on src/main/java/org/apache/sling/dynamicinclude/api/IncludeGenerator.java:
```diff
@@ -17,13 +17,15 @@
     String getType();
 
-    String getInclude(String url);
+    String getInclude(SlingHttpServletRequest request,String url);
```
I think it's confusing to pass in the request and a string named `url` I see that the `url` parameter is based on the sling request, so maybe we can have a different approach here.

santiagozky replied:
the issue here is that the url is not just taken from the request as it is. It is previously checked to do things like removing the query part, map it,encode, etc (in https://github.com/apache/sling-org-apache-sling-dynamic-include/blob/master/src/main/java/org/apache/sling/dynamicinclude/IncludeTagFilter.java#L149),.
we could remove the url parameter, but that means any IncludeGenerator would be responsible for doing that

rombert replied:
I think that would be ok, granted that we:

- find a better name for it ( normalizedUrl?)
- add better documentation to the interface and the method, explaining why we pass in an apparently redundant parameter

rombert reviewed (changes requested):
Sorry for the delay on my side, comments in the review.

santiagozky commented:
Sorry for the delay @rombert. I was too busy and this feltout of priority for me.  I've added the built-in generators inthe configuration label and added the method javadoc

rombert reviewed (approved):
LGTM, thanks.

santiagozky commented:
thank you @rombert
